### PR TITLE
a major refactor of example/image-classification

### PR DESCRIPTION
**Pull request by mli**
this is a major refactor of image-classification. changes include:

1. moved common codes such as argument, data iterator and training into common/
2. moved symbol definitions into symbol/
3. added scripts to prepare data and download pretrained models
4. added score.py and fine-tune.py
5. updated README
6. added resnet from https://github.com/tornadomeet/ResNet

all *.py in the root fold are tested manually.

some TODOs

1. refactor the R part
2. instructions to enable mirror
3. double check benchmark.py on multiple machines
4. a test/ fold with scripts to test this example

### CODE REVIEW

**Comment by piiswrong** [comment]
We need some test for examples. Do you want to merge this first or add testing together with this pr?

File: example/image-classification/common/fit.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+    # devices for training
+    devs = mx.cpu() if args.gpus is None or args.gpus is '' else [
+        mx.gpu(int(i)) for i in args.gpus.split(',')]
+
+    # learning rate
+    lr, lr_scheduler = _get_lr_scheduler(args, kv)
+
+    # create model
+    model = mx.model.FeedForward(
```
Comment (piiswrong) [inline example/image-classification/common/fit.py]: need to change to module in the future. I'm planning to add deprecation warnings for model with NNVM update

File: example/image-classification/mnist_bucket.py [hunk 2]
```diff
@@ -1,95 +0,0 @@
...
-    args = train_mnist.parse_args()
-
-    args.network = 'mlp'
-    net = train_mnist.get_mlp()
-
-    buckets = [1, 2, 3]
-    def symbol_generator(key):
-        return net # all the symbols are the same for all bucket entries
-
-    train_mnist.train_model.fit(args, symbol_generator, get_iterator(buckets))
```
Comment (piiswrong) [inline example/image-classification/mnist_bucket.py]: moving the files may cause confusion for users. Leave a note somewhere?

File: example/image-classification/score.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    args.model, os.path.join(dir_path, 'model'))
+
+# create module
+sym, arg_params, aux_params = mx.model.load_checkpoint(prefix, epoch)
+if args.gpus == '':
+    devs = mx.cpu()
+else:
+    devs = [mx.gpu(int(i)) for i in args.gpus.split(',')]
+
+mod = mx.mod.Module(symbol=sym, context=devs)
```
Comment (piiswrong) [inline example/image-classification/score.py]: score uses module while fit uses model can lead to confusions.

File: example/image-classification/symbol/resnet.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        Ouput size of symbol
+    dataset : str
+        Dataset type, only cifar10 and imagenet supports
+    workspace : int
+        Workspace used in convolution operator
+    """
+    num_unit = len(units)
+    assert(num_unit == num_stages)
+    data = mx.sym.Variable(name='data')
+    data = mx.sym.BatchNorm(data=data, fix_gamma=True, eps=2e-5, momentum=bn_mom, name='bn_data')
```
Comment (piiswrong) [inline example/image-classification/symbol/resnet.py]: Is data normalization method different for different symbols? How is this handled in score and train?

File: example/image-classification/train_cifar10_mirroring.py [hunk 5]
```diff
@@ -1,132 +0,0 @@
-#################################################################################
```
Comment (piiswrong) [inline example/image-classification/train_cifar10_mirroring.py]: why remove this?

File: example/image-classification/common/fit.py (hunk 1 above)
Comment (mli) [inline example/image-classification/common/fit.py]: yup, on the TODO list, but i don't want to change too much codes in this PR

File: example/image-classification/mnist_bucket.py (hunk 2 above)
Comment (mli) [inline example/image-classification/mnist_bucket.py]: mentioned in readme.md. but we are going to remove a bunch of folds in example/ anyway.

File: example/image-classification/symbol/resnet.py (hunk 4 above)
Comment (mli) [inline example/image-classification/symbol/resnet.py]: what do you mean?

File: example/image-classification/train_cifar10_mirroring.py (hunk 5 above)
Comment (mli) [inline example/image-classification/train_cifar10_mirroring.py]: added mirroring in readme

File: example/image-classification/symbol/resnet.py (hunk 4 above)
Comment (piiswrong) [inline example/image-classification/symbol/resnet.py]: So here in resnet symbol there is a BatchNorm in the begining that does mean substraction and std normalization.
The inception symbols don't have this and use ImageRecordIter(mean_r=xxx) instead. Is the script handling this?

**Comment by piiswrong** [comment]
LGTM

File: example/image-classification/test_score.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+def test_imagenet1k_resnet(args):
+    models = ['imagenet1k-resnet-34',
+              'imagenet1k-resnet-50',
+              'imagenet1k-resnet-101',
+              'imagenet1k-resnet-152']
+    accs = [.72, .75, .765, .76]
+    for (m, g) in zip(models, accs):
+        acc = mx.metric.create('acc')
+        (speed,) = score(model=m, data_val='data/val-5k-256.rec',
+                         rgb_mean='0,0,0', metrics=acc, **vars(args))
```
Comment (piiswrong) [inline example/image-classification/test_score.py]: @tornadomeet 
All our pretrained models except resnet use mean substract for input.
Could you update the pretrained resnet models (add 123.68,116.779,103.939 to moving_mean) and your resnet repo so that we can use the same script to score all models?

Comment (tornadomeet) [inline example/image-classification/test_score.py]: @piiswrong actually we can do this, but i think we'd better not do that and add one parameters like ```rgb_mean``` . because no need mean value is more convenient than the value like (123.68,116.779,103.939) and this is original considered for resnet training on purpose.
another, in some training project(such as face-related model ), we usually use mean=[127.5, 127.5, 127.5], and this is also a different mean value. or on some other field, the mean value may even own a big difference than 127.  
so we can give write corresponding code for some famous model, and just mention it in the readme file.

Comment (mli) [inline example/image-classification/test_score.py]: there is a `--rgb-mean` option

Comment (mli) [inline example/image-classification/test_score.py]: but it would be great if we have a default rgb_mean for all pre-trained models

Comment (piiswrong) [inline example/image-classification/test_score.py]: The goal is to be consistent. Currently all other models except resnet need mean substraction. In the short term we should modify resnet to be consistent with the other models. In the long term we should retrain all the other models to make a batchnorm layer as first layer standard.

Note that mean substraction doesn't exclude using bn on the first layer, it will have no effect. This is the best solution I can think of that minimize confusion.

**Comment by thirdwing** [comment]
I will update the R part.

BTW, is it possible to move the R testing to jenkins?

**Comment by piiswrong** [comment]
Its on travis because @tqchen said osx-R is more "strict" than linux-R and
jenkins doesn't have osx. If you think its not an issue we can move it to
jenkins

**Comment by thirdwing** [comment]
We can have both. The reason I want to move it to jenkins is that R test takes quite a long time.

To meet the time limit on Travis-CI, some examples and test are disabled.
